File: src/SOP8/Utility/dnsExcelReport/Writer/ExcelWriter.cs
using NPOI.HPSF;
using NPOI.HSSF.UserModel;
using NPOI.SS.UserModel;
using System.Collections.Generic;
using System.IO;
using System;
using System.Collections;

namespace dnsExcelReport.Writer
{
    using Models;

    public abstract class ExcelWriter
    {
        public enum TableHeaderMode { Left = 1, Middle = 2, Right = 4 };
        public enum TableBodyRow { Top = 1, Middle = 2, Bottom = 4 };

        public ExcelWriter()
        {
        }

        public byte[] Run(out string strErrorMessage)
        {
            try
            {
                strErrorMessage = null;
                HSSFWorkbook workbook = MakeWorkbook();

                if (workbook == null)
                    return null;

                ICollection<SheetData> sheetDatas = ReadSheetDatas(workbook, out strErrorMessage);

                if (sheetDatas == null)
                {
                    System.Diagnostics.Trace.WriteLine(strErrorMessage);
                    return null;
                }

                WriteSheetDatas(workbook, sheetDatas);

                byte[] bytes = null;

                using (MemoryStream stream = new MemoryStream())
                {
                    workbook.Write(stream);
                    bytes = stream.ToArray();
                }

                workbook.Close();
                return bytes;
            }
            catch (System.Exception e)
            {
                System.Diagnostics.Trace.WriteLine(e.Message);
                strErrorMessage = e.Message;
            }

            return null;
        }

        protected void WriteSheetDatas(HSSFWorkbook workbook, ICollection<SheetData> sheetDatas)
        {
            foreach (SheetData sheetData in sheetDatas)
            {
                ISheet sheet = workbook.CreateSheet(sheetData.SheetName);

                if (sheet == null)
                    return;

                int nextRowIndex = WritePrev(sheet, workbook);

                IRow row = sheet.CreateRow(nextRowIndex);

                if (sheetData.TitleRowHeight != null)
                    row.HeightInPoints = (int)sheetData.TitleRowHeight;

                int min, max;

                if (GetMinMax(sheetData.Titles, out max, out min) == false)
                    continue;

                string strTitle;

                for (int i = min; i <= max; i++)
                {
                    if (sheetData.Titles.TryGetValue(i, out strTitle))
                    {
                        ICell cell = row.CreateCell(i);

                        ICellStyle style;

                        if (sheetData.TitleStyles.TryGetValue(i, out style))
                            cell.CellStyle = style;

                        if (cell != null && strTitle != null)
                            cell.SetCellValue(strTitle);
                    }
                }

                List<string> values;
                // Key : Column Index
                Dictionary<int, IRow> dicColumnRows = new Dictionary<int, IRow>();

                for (int i = min; i <= max; i++)
                {
                    if (sheetData.ColumnDatas.TryGetValue(i, out values))
                    {
                        int nValueCount = values.Count;

                        for (int j = 0; j < nValueCount; j++)
                        {
                            if (dicColumnRows.TryGetValue(j, out row) == false)
                            {
                                row = sheet.CreateRow(j + 1 + nextRowIndex);
                                dicColumnRows[j] = row;

                                int rowHeight;

                                if (sheetData.RowHeight.TryGetValue(j, out rowHeight))
                                    row.HeightInPoints = rowHeight;
                            }

                            string str = values[j];
                            ICell cell = row.CreateCell(i);

                            Dictionary<int, ICellStyle> dicStyles;

                            if (sheetData.CellStyles.TryGetValue(i, out dicStyles))
                            {
                                ICellStyle style;

                                if (dicStyles.TryGetValue(j, out style))
                                    cell.CellStyle = style;
                            }

                            if (cell != null && str != null)
                                cell.SetCellValue(str);
                        }
                    }
                }

                WritePost(sheet, workbook);
            }
        }

        protected virtual int WritePrev(ISheet sheet, HSSFWorkbook workbook)
        {
            return 0;
        }

        protected virtual void WritePost(ISheet sheet, HSSFWorkbook workbook)
        {
        }

        protected bool GetMinMax(Dictionary<int, string> dicTitles, out int max, out int min)
        {
            max = -1;
            min = 1;

            foreach (KeyValuePair<int, string> pair in dicTitles)
            {
                if (min > max)
                {
                    min = max = pair.Key;
                }
                else
                {
                    if (min > pair.Key)
                        min = pair.Key;

                    if (max < pair.Key)
                        max = pair.Key;
                }
            }

            return min <= max;
        }

        private HSSFWorkbook MakeWorkbook()
        {
            string strCompany = "유엔이";

            if (strCompany == null)
                strCompany = "";

            HSSFWorkbook hssfworkbook = new HSSFWorkbook(/*stream*/);

            DocumentSummaryInformation dsi = PropertySetFactory.CreateDocumentSummaryInformation();
            dsi.Company = strCompany;
            hssfworkbook.DocumentSummaryInformation = dsi;

            //create a entry of SummaryInformation
            SummaryInformation si = PropertySetFactory.CreateSummaryInformation();
            si.Subject = GetSubject();
            hssfworkbook.SummaryInformation = si;

            return hssfworkbook;
        }

        protected abstract ICollection<SheetData> ReadSheetDatas(HSSFWorkbook workbook, out string strErrorMessage);
        protected abstract string GetSubject();

        public virtual string GetFileName()
        {
            return GetFileName(GetSubject());
        }

        protected string GetFileName(string strTag)
        {
            DateTime dtNow = DateTime.Now;
            return string.Format("{0}_{1}{2:00}{3:00}_{4:00}{5:00}{6:00}.xls", strTag, dtNow.Year, dtNow.Month, dtNow.Day, dtNow.Hour, dtNow.Minute, dtNow.Second);
        }

        protected static bool IsLeft(int mode)
        {
            if ((mode & (int)TableHeaderMode.Left) == (int)TableHeaderMode.Left)
                return true;

            return false;
        }

        protected static bool IsHMiddle(int mode)
        {
            if ((mode & (int)TableHeaderMode.Middle) == (int)TableHeaderMode.Middle)
                return true;

            return false;
        }

        protected static bool IsRight(int mode)
        {
            if ((mode & (int)TableHeaderMode.Right) == (int)TableHeaderMode.Right)
                return true;

            return false;
        }

        protected static bool IsTop(int mode)
        {
            if ((mode & (int)TableBodyRow.Top) == (int)TableBodyRow.Top)
                return true;

            return false;
        }

        protected static bool IsVMiddle(int mode)
        {
            if ((mode & (int)TableBodyRow.Middle) == (int)TableBodyRow.Middle)
                return true;

            return false;
        }

        protected static bool IsBottom(int mode)
        {
            if ((mode & (int)TableBodyRow.Bottom) == (int)TableBodyRow.Bottom)
                return true;

            return false;
        }
    }
}

File: src/SOP8/Egovframework/Base/base-utility/src/main/java/egovframework/base/excelreport/writer/ExcelWriter.java
package egovframework.base.excelreport.writer;

import egovframework.base.excelreport.models.SheetData;
import org.apache.poi.hpsf.DocumentSummaryInformation;
import org.apache.poi.hpsf.SummaryInformation;
import org.apache.poi.hpsf.PropertySetFactory;
import org.apache.poi.hssf.usermodel.HSSFWorkbook;
import org.apache.poi.ss.usermodel.*;

import java.io.ByteArrayOutputStream;
import java.util.*;

public abstract class ExcelWriter {

    public enum TableHeaderMode {
        Left(1), Middle(2), Right(4);
        private final int value;
        TableHeaderMode(int value) { this.value = value; }
        public int getValue() { return value; }
    }

    public enum TableBodyRow {
        Top(1), Middle(2), Bottom(4);
        private final int value;
        TableBodyRow(int value) { this.value = value; }
        public int getValue() { return value; }
    }

    public ExcelWriter() {}

    public byte[] run(StringBuilder errorMessage) {
        try {
            errorMessage.setLength(0);
            HSSFWorkbook workbook = makeWorkbook();

            if (workbook == null) {
                return null;
            }

            Collection<SheetData> sheetDatas = readSheetDatas(workbook, errorMessage);

            if (sheetDatas == null) {
                System.err.println(errorMessage);
                return null;
            }

            writeSheetDatas(workbook, sheetDatas);

            byte[] bytes;
            try (ByteArrayOutputStream bos = new ByteArrayOutputStream()) {
                workbook.write(bos);
                bytes = bos.toByteArray();
            }
            workbook.close();
            return bytes;

        } catch (Exception e) {
            errorMessage.append(e.getMessage());
            return null;
        }
    }

    protected void writeSheetDatas(HSSFWorkbook workbook, Collection<SheetData> sheetDatas) {
        for (SheetData sheetData : sheetDatas) {
            Sheet sheet = workbook.createSheet(sheetData.getSheetName());

            if (sheet == null) continue;

            int nextRowIndex = writePrev(sheet, workbook);

            Row row = sheet.createRow(nextRowIndex);
            if (sheetData.getTitleRowHeight() != null) {
                row.setHeightInPoints(sheetData.getTitleRowHeight());
            }

            int min, max;
            int[] minMax = getMinMax(sheetData.getTitles());
            if (minMax == null) continue;

            min = minMax[0];
            max = minMax[1];

            for (int i = min; i <= max; i++) {
                String strTitle = sheetData.getTitles().get(i);
                if (strTitle != null) {
                    Cell cell = row.createCell(i);

                    CellStyle style = sheetData.getTitleStyles().get(i);
                    if (style != null) {
                        cell.setCellStyle(style);
                    }
                    cell.setCellValue(strTitle);
                }
            }

            Map<Integer, Row> rowMap = new HashMap<>();
            for (int i = min; i <= max; i++) {
                List<String> values = sheetData.getColumnDatas().get(i);
                if (values != null) {
                    for (int j = 0; j < values.size(); j++) {
                        Row dataRow = rowMap.computeIfAbsent(j, k -> sheet.createRow(k + 1 + nextRowIndex));

                        Integer rowHeight = sheetData.getRowHeight().get(j);
                        if (rowHeight != null) {
                            dataRow.setHeightInPoints(rowHeight);
                        }

                        Cell cell = dataRow.createCell(i);
                        String value = values.get(j);

                        Map<Integer, CellStyle> styleMap = sheetData.getCellStyles().get(i);
                        if (styleMap != null) {
                            CellStyle style = styleMap.get(j);
                            if (style != null) {
                                cell.setCellStyle(style);
                            }
                        }
                        if (value != null) {
                            cell.setCellValue(value);
                        }
                    }
                }
            }
            writePost(sheet, workbook);
        }
    }

    protected int[] getMinMax(Map<Integer, String> dicTitles) {
        int max = -1;
        int min = Integer.MAX_VALUE;

        for (Integer key : dicTitles.keySet()) {
            if (min > max) {
                min = max = key;
            } else {
                if (min > key) min = key;
                if (max < key) max = key;
            }
        }

        return (min <= max) ? new int[]{min, max} : null;
    }

    private HSSFWorkbook makeWorkbook() {
        String company = "유엔이";
        HSSFWorkbook workbook = new HSSFWorkbook();

        try {
            DocumentSummaryInformation dsi = PropertySetFactory.newDocumentSummaryInformation();
            dsi.setCompany(company);
            workbook.createInformationProperties();
            workbook.getDocumentSummaryInformation().setCompany(company);

            SummaryInformation si = workbook.getSummaryInformation();
            si.setSubject(getSubject());
        } catch (Exception e) {
            e.printStackTrace();
        }
        return workbook;
    }

    protected abstract Collection<SheetData> readSheetDatas(HSSFWorkbook workbook, StringBuilder errorMessage);
    protected abstract String getSubject();

    public String getFileName() {
        return getFileName(getSubject());
    }

    protected String getFileName(String strTag) {
        Calendar cal = Calendar.getInstance();
        return String.format("%s_%d%02d%02d_%02d%02d%02d.xls",
                strTag,
                cal.get(Calendar.YEAR),
                cal.get(Calendar.MONTH) + 1,
                cal.get(Calendar.DAY_OF_MONTH),
                cal.get(Calendar.HOUR_OF_DAY),
                cal.get(Calendar.MINUTE),
                cal.get(Calendar.SECOND));
    }

    protected int writePrev(Sheet sheet, HSSFWorkbook workbook) {
        return 0;
    }

    protected void writePost(Sheet sheet, HSSFWorkbook workbook) {}

    // Utility methods for alignment flags
    protected static boolean isLeft(int mode) { return (mode & TableHeaderMode.Left.getValue()) == TableHeaderMode.Left.getValue(); }
    protected static boolean isHMiddle(int mode) { return (mode & TableHeaderMode.Middle.getValue()) == TableHeaderMode.Middle.getValue(); }
    protected static boolean isRight(int mode) { return (mode & TableHeaderMode.Right.getValue()) == TableHeaderMode.Right.getValue(); }
    protected static boolean isTop(int mode) { return (mode & TableBodyRow.Top.getValue()) == TableBodyRow.Top.getValue(); }
    protected static boolean isVMiddle(int mode) { return (mode & TableBodyRow.Middle.getValue()) == TableBodyRow.Middle.getValue(); }
    protected static boolean isBottom(int mode) { return (mode & TableBodyRow.Bottom.getValue()) == TableBodyRow.Bottom.getValue(); }
}